### ISSUE TITLE
how to solve GDS error

### ISSUE THREAD

**Issue by ksm9413** [developer-onizuka/gpudirect_storage_1.0#2]
Hello developer-onizuka,

i can install gds according to your guide thank you very much. I was able to get the same gdscheck output:
But while testing gds I got an error

root@syseng-desktop:/usr/local/cuda-11.4/gds/tools# ./gdscheck -p
 GDS release version: 1.0.0.82
 nvidia_fs version:  2.7 libcufile version: 2.4
 ============
 ENVIRONMENT:
 ============
 =====================
 DRIVER CONFIGURATION:
 =====================
 NVMe               : Supported
 NVMeOF             : Unsupported
 SCSI               : Unsupported
 ScaleFlux CSD      : Unsupported
 NVMesh             : Unsupported
 DDN EXAScaler      : Unsupported
 IBM Spectrum Scale : Unsupported
 NFS                : Unsupported
 WekaFS             : Unsupported
 Userspace RDMA     : Unsupported
 --Mellanox PeerDirect : Enabled
 --rdma library        : Not Loaded (libcufile_rdma.so)
 --rdma devices        : Not configured
 --rdma_device_status  : Up: 0 Down: 0
 =====================
 CUFILE CONFIGURATION:
 =====================
 properties.use_compat_mode : true
 properties.gds_rdma_write_support : true
 properties.use_poll_mode : false
 properties.poll_mode_max_size_kb : 4
 properties.max_batch_io_timeout_msecs : 5
 properties.max_direct_io_size_kb : 16384
 properties.max_device_cache_size_kb : 131072
 properties.max_device_pinned_mem_size_kb : 33554432
 properties.posix_pool_slab_size_kb : 4 1024 16384 
 properties.posix_pool_slab_count : 128 64 32 
 properties.rdma_peer_affinity_policy : RoundRobin
 properties.rdma_dynamic_routing : 0
 fs.generic.posix_unaligned_writes : false
 fs.lustre.posix_gds_min_kb: 0
 fs.weka.rdma_write_support: false
 profile.nvtx : false
 profile.cufile_stats : 0
 miscellaneous.api_check_aggressive : false
 =========
 GPU INFO:
 =========
 GPU index 0 Quadro P4000 bar:1 bar size (MiB):256 supports GDS
 ==============
 PLATFORM INFO:
 ==============
 IOMMU: disabled
 Platform verification succeeded

####################
dmesg

[ 6.393409] nvidia_fs: Initializing nvfs driver module
[ 6.393410] nvidia_fs: registered correctly with major number 239
[ 6.393586] nvidia-fs⚠️ error retrieving numa node for device 0000:01:00.0
[ 6.393588] nvidia-fs⚠️ error retrieving numa node for device 0000:00:1f.6
[ 6.393590] nvidia-fs⚠️ error retrieving numa node for device 0000:01:00.0
[ 6.393592] nvidia-fs⚠️ error retrieving numa node for device 0000:02:00.0

####################
dmesg after running gdsio

[ 690.004369] EXT4-fs (nvme0n1): recovery complete
[ 690.004896] EXT4-fs (nvme0n1): mounted filesystem with ordered data mode. Opts: data=ordered
[ 748.642116] nvme nvme0: controller is down; will reset: CSTS=0x3, PCI_STATUS=0x2010
[ 748.705906] nvme nvme0: Shutdown timeout set to 8 seconds
[ 748.758419] nvme nvme0: 12/0/0 default/read/poll queues
[ 779.362181] nvme nvme0: I/O 1004 QID 1 timeout, disable controller
[ 779.394769] blk_update_request: I/O error, dev nvme0n1, sector 22699520 op 0x1:(WRITE) flags 0x8800 phys_seg 1 prio class 0
[ 779.394781] blk_update_request: I/O error, dev nvme0n1, sector 22696960 op 0x1:(WRITE) flags 0xc800 phys_seg 2 prio class 0
[ 779.394788] blk_update_request: I/O error, dev nvme0n1, sector 22694400 op 0x1:(WRITE) flags 0xc800 phys_seg 1 prio class 0
[ 779.394793] blk_update_request: I/O error, dev nvme0n1, sector 22691840 op 0x1:(WRITE) flags 0xc800 phys_seg 4 prio class 0
[ 779.394851] nvme nvme0: failed to mark controller live state
[ 779.394855] nvme nvme0: Removing after probe failure status: -19
[ 779.394873] nvidia-fs:write IO failed :-5
[ 779.418191] blk_update_request: I/O error, dev nvme0n1, sector 486801416 op 0x1:(WRITE) flags 0x800 phys_seg 8 prio class 0
[ 779.418226] Aborting journal on device nvme0n1-8.
[ 779.418242] Buffer I/O error on dev nvme0n1, logical block 60850176, lost sync page write
[ 779.418248] JBD2: Error -5 detected when updating journal superblock for nvme0n1-8.
[ 779.447856] nvme nvme0: failed to set APST feature (-19)

#####################
gdsio returrn message

root@syseng-desktop:/usr/local/cuda-11.4/gds/tools# ./gdsio -f /mnt/test10G -d 0 -n 0 -w 1 -s 10G -x 0 -I 1 -T 10 -i 4096K
Error: IO failed stopping traffic, fd :27 ret:-5 errno :1
io failed :ret :-5 errno :1, file offset :0, block size :4194304

######################
cufile.log

16-07-2021 17:11:30:732 [pid=1761 tid=1761] ERROR cufio-udev:441 no devices found by udev enumeration for pci class: infiniband
16-07-2021 17:11:30:751 [pid=1761 tid=1761] ERROR cufio-udev:441 no devices found by udev enumeration for pci class: infiniband
16-07-2021 17:11:30:752 [pid=1761 tid=1761] ERROR cufio-udev:441 no devices found by udev enumeration for pci class: infiniband
16-07-2021 17:21:07:820 [pid=1939 tid=1939] ERROR cufio-udev:441 no devices found by udev enumeration for pci class: infiniband
16-07-2021 17:21:07:821 [pid=1939 tid=1939] ERROR cufio-udev:441 no devices found by udev enumeration for pci class: infiniband
16-07-2021 17:21:07:822 [pid=1939 tid=1939] ERROR cufio-udev:441 no devices found by udev enumeration for pci class: infiniband
16-07-2021 17:22:10:325 [pid=1939 tid=1943] ERROR 0:1405 IOCTL failed io-type 1 ret -5 expected 4194304 gpu_page_offset 0

I'm having trouble solving this problem
I am running Ubuntu 20.04.2 LTS, kernel version 5.4.0-42-generic.
The CPU is i7-10700k, and the nvme is Samsung 980 m.2 model.

I would really appreciate it if you could let me know what is causing this issue

**Comment by developer-onizuka** [comment]
Hi ksm9413,

Thanks for your comment.
In my experience, the NVMe which has cache DRAM might not be used for GDS. 
Actually, I could not use the Crucial NVMe p5 serise (CT500P5SSD8JP) for GDS.
https://www.crucial.com/products/ssd/crucial-p5-ssd

If your samsung NVMe has cache DRAM, it might not also be used as GDS like mine.
I believe that the NVMe whoes read throughput is around 2000MB/s has no cache. 
Try to exchange to other cheap NVMe which has no cache DRAMs.

Best,

**Comment by developer-onizuka** [comment]
I also have Samsung's NVMe (P/N: MZVLW128HEGR-000L1).
I tried GDS about it and then I also got the following dmesg.

$ dmesg |grep nvidia
[    2.204709] nvidia_fs: Initializing nvfs driver module
[    2.204711] nvidia_fs: registered correctly with major number 239
[    2.205073] nvidia-fs:warning: error retrieving numa node for device 0000:04:00.0 
[    2.205075] nvidia-fs:warning: error retrieving numa node for device 0000:01:00.0 
[    2.205079] nvidia-fs:warning: error retrieving numa node for device 0000:04:00.0 
[    2.205082] nvidia-fs:warning: error retrieving numa node for device 0000:06:00.0 

I actually failed at GDS write pattern. But I can run GDS read pattern as followings:

$ gdsio -f /mnt/test10G -d 0 -n 0 -w 1 -s 10G -x 0 -I 0 -T 10 -i 256K
IoType: READ XferType: GPUD Threads: 1 DataSetSize: 17141760/10485760(KiB) IOSize: 256(KiB) Throughput: 1.650979 GiB/sec, Avg_Latency: 147.865965 usecs ops: 66960 total_time 9.901797 secs

See also URL below:
https://github.com/developer-onizuka/gpudirect_storage_1.0_VirtualMachine

This is a result I created GDS environment on virtual machine.

Best,
